pandas/_libs/src/vendored/numpy/datetime/np_datetime.c

```diff
@@ -711,355 +711,120 @@ void pandas_datetime_to_datetimestruct(npy_datetime dt, NPY_DATETIMEUNIT base,
-        out->days = -frac / 86400LL + 1;
-        frac += 86400LL * out->days;
-      } else {
-        frac = -frac;
-      }
-    } else {
-      sign = 1;
-      out->days = 0;
-    }
-
-    if (frac >= 86400) {
-      out->days += frac / 86400LL;
-      frac -= out->days * 86400LL;
-    }
-
-    if (frac >= 3600) {
-      out->hrs = frac / 3600LL;
-      frac -= out->hrs * 3600LL;
-    } else {
-      out->hrs = 0;
-    }
-
-    if (frac >= 60) {
-      out->min = frac / 60LL;
-      frac -= out->min * 60LL;
-    } else {
-      out->min = 0;
-    }
-
-    if (frac >= 0) {
-      out->sec = frac;
-      frac -= out->sec;
-    } else {
-      out->sec = 0;
-    }
-
-    sfrac = (out->hrs * 3600LL + out->min * 60LL + out->sec) * per_sec;
-
-    if (sign < 0)
-      out->days = -out->days;
-
-    ifrac = td - (out->days * per_day + sfrac);
-
-    if (ifrac != 0) {
-      out->ms = ifrac / 1000LL;
-      ifrac -= out->ms * 1000LL;
-      out->us = ifrac / 1L;
-      ifrac -= out->us * 1L;
-      out->ns = ifrac;
-    } else {
-      out->ms = 0;
-      out->us = 0;
-      out->ns = 0;
-    }
+  case (NPY_FR_D):
+    out->days = td;
     break;
-
+  case NPY_FR_h:
+    out->days = td / 24LL;
+    td -= out->days * 24LL;
+    out->hrs = td;
+    break;
+  case NPY_FR_m:
+    out->days = td / 1440LL;
+    td -= out->days * 1440LL;
+    out->hrs = td / 60LL;
+    td -= out->hrs * 60LL;
+    out->min = td;
+    break;
+  case NPY_FR_s:
   case NPY_FR_ms:
-
-    per_day = 86400000LL;
-    per_sec = 1000LL;
-
-    // put frac in seconds
-    if (td < 0 && td % per_sec != 0)
-      frac = td / per_sec - 1;
-    else
-      frac = td / per_sec;
-
-    if (frac < 0) {
-      sign = -1;
-
-      // even fraction
-      if ((-frac % 86400LL) != 0) {
-        out->days = -frac / 86400LL + 1;
-        frac += 86400LL * out->days;
-      } else {
-        frac = -frac;
-      }
-    } else {
-      sign = 1;
-      out->days = 0;
-    }
-
-    if (frac >= 86400) {
-      out->days += frac / 86400LL;
-      frac -= out->days * 86400LL;
-    }
-
-    if (frac >= 3600) {
-      out->hrs = frac / 3600LL;
-      frac -= out->hrs * 3600LL;
-    } else {
-      out->hrs = 0;
-    }
-
-    if (frac >= 60) {
-      out->min = frac / 60LL;
-      frac -= out->min * 60LL;
-    } else {
-      out->min = 0;
-    }
-
-    if (frac >= 0) {
-      out->sec = frac;
-      frac -= out->sec;
-    } else {
-      out->sec = 0;
-    }
-
-    sfrac = (out->hrs * 3600LL + out->min * 60LL + out->sec) * per_sec;
-
-    if (sign < 0)
-      out->days = -out->days;
-
-    ifrac = td - (out->days * per_day + sfrac);
-
-    if (ifrac != 0) {
-      out->ms = ifrac;
-      out->us = 0;
-      out->ns = 0;
+  case NPY_FR_us:
+  case NPY_FR_ns: {
+    const npy_int64 sec_per_day = 86400;
+    npy_int64 per_sec;
+    if (base == NPY_FR_s) {
+      per_sec = 1;
+    } else if (base == NPY_FR_ms) {
+      per_sec = 1000;
+    } else if (base == NPY_FR_us) {
+      per_sec = 1000000;
```

Comment:
overlap with tslibs.dtypes.periods_per_second.  probably not worth the hassle of de-duplicating

Reply:
If we wanted to we could make that an inline function like:

```c
inline npy_int64 per_sec_multiplier(NPY_DATETIMEUNIT base):
    switch base:
    case NPY_FR_s:
        return 1;
    case NPY_FR_ms:
        return 1000;
    ...
```